### ISSUE TITLE
Structured log data from `stderr` using JSON?

### ISSUE THREAD

**Issue by ioquatix** [systemd/systemd#19683]
I have been reading about the log formats available and the various ways to log data from a program. Thank you for the existing documentation.

I would like to propose that we add support for structured logging using newline delimited JSON via `stderr`.

Programs could output JSON to `stderr` and have the internal structure picked up and formatted by the systemd journal.

I would suggest we can sniff JSON by checking the first and last character in the message is `{` and `}` respectively.

This approach to logging is supported by a lot of log aggregation services so it seems like a nice way to have consistency across a wide range of platforms.

**Comment by poettering** [comment]
Not sure I follow. I am not aware of any programs that output JSON via stderr. That'd be a new thing. But if it's a new thing why not user a more appropriate protocol?

Moreover, journald is not a generic JSON processor, but stores data only indexed by the simple env-like block structure it uses for log recrods, hence accepting jSON is a bit useless, since we wouldn't make use of it.

so what i your intetnion here? do you have some other log processor behind journald? if so, that's fine, journald will propagate the data as-is, so you can just send json data if you like.

there are a myriad of packages that can generate native journal log entries. Which languages do you have in mind where that' still missing?

**Comment by ioquatix** [comment]
@poettering thanks for your quick reply I appreciate you giving your time to discuss this issue.

> output JSON via stderr. That'd be a new thing

Quite a lot of logging aggregation services use JSON, e.g. Splunk, Fluentd, etc. So it's not that uncommon for applications generate log output using JSON. The benefit is that it's very easy to implement, a lot of applications emit structured events with key-value pairs - sometimes nested. Whether it goes via `stdout` or `stderr` depends on the design of the program. But for diagnostics, I would normally suggest `stderr`.

> Moreover, journald is not a generic JSON processor, but stores data only indexed by the simple env-like block structure it uses for log recrods, hence accepting jSON is a bit useless, since we wouldn't make use of it.

So, I understand that, and it makes sense to me. I think there are two sides to this:

## Data Ingestion

Currently, we can emit JSON, and it can be processed. So maybe a non-issue. But it would be nice if we could detect a content type/format for logging. In some logging programs, I check whether a TTY is attached, and if so, use a more human-oriented format for logs. If the output is going to a file or some other non-TTY device, I'll format it in a more structured way, e.g. JSON.

As you say, journald can accept any format. However, what I did wonder is whether we should configure in the systemd unit file, the output format content type, e.g. JSON. It might be a nice way to provide a little bit more structure to the output format. Other valid formats might be the existing RFC3164 format, etc.

Another option I was thinking about, similar to how you handle `NOTIFY_SOCKET`, is to implement some log format environment variable, e.g. `LOG_FORMAT=RFC3164` or `LOG_FORMAT=JSON` etc. This could be controlled by the systemd unit configuration.

## Data Presentation

The log data, as you say, is currently a simple key-value storage. Regarding JSON, we could define some schema, e.g.

```
{severity: ..., subject: ..., pid: ..., message: ...}
```

If we defined a short schema like this, `journalctl` could interpret and present the JSON log messages in a way that was easier for human consumption. It might also serve as a starting point for log indexing and querying. While I don't know how far would make sense to implement it with journalctl, I could certainly imagine taking those log entires and ingesting them into a dedicated processor.

The way I see it, is that I'd like to embrace journalctl as much as possible, so from my POV, I'd like the user experience to be as good as possible. I'd rather emit structured events so that we have the data available to provide that user experience in a logical way i.e. filtering/querying by keys/values, presenting log messages rather than JSON blobs.

I'd be perfectly happy to have this as a separate service, but I really think journalctl is excellent and if it makes sense to you, maybe there is an opportunity here to exploit a simpler way to have structured event logs with a great UX.

**Comment by DemiMarie** [comment]
Would it be possible to use `systemd-journald`’s native protocol to emit structured data directly?

**Comment by ioquatix** [comment]
@DemiMarie I'd be fine with that but we need to know the content type that is expected by the app. I think if that's the expectation, we need something like `LOG_FORMAT=JOURNAL` so that I know to output that format without the user providing some option, etc.

**Comment by DemiMarie** [comment]
> @DemiMarie I'd be fine with that, but it's unlikely to be a standard format supported by log aggregation services. In order to do this, we need to know the content type that is expected by the app. I think if that's the expectation, we need something like `LOG_FORMAT=JOURNAL` so that I know to output that format without the user providing some option, etc.

That can be set via an environment variable in the unit file, but it should probably be set for all units.

**Comment by ioquatix** [comment]
@DemiMarie Yes, that makes sense to me. I suspect the most general format is JSON, but a journal specific format would also work. In that case, I suggest we work on both?

In the unit file, could be:

```
JournalFormat=systemd
JournalFormat=json
```

**Comment by ioquatix** [comment]
Just wondering how we make progress with this issue.

- Specify a logging format as an environment variable? (so we know to output journald specific format).
- Support and/or default to NDJSON (standardised across all major log aggregation services that I'm aware of, e.g. Splunk, Datadog, etc).
- Enable some kind of JournalFormat configuration in the unit file.

Thanks!

**Comment by ivoanjo** [comment]
Hey 👋. First of all, thanks for systemd! I'm pretty happy with it, and I quite like how it centralized a whole big ball of unix stuff I previously quite struggled with.

I was discussing this with @ioquatix and definitely I think we have the onus of proving that JSON logging is a thing :)

So here's me taking a stab at it, hopefully it's useful.

Library side:

* For Java, Log4j, the famous ~RPC~ logging library includes support for JSON logging ([1](https://logging.apache.org/log4j/2.x/manual/layouts.html#JSONLayout), [2](https://logging.apache.org/log4j/2.x/manual/json-template-layout.html)). I previously worked at AWS and my team's services all configured the logger using this layout. There's also an alternative JSON pattern from the logstash project: [3](https://github.com/logstash/log4j-jsonevent-layout) .
* For Ruby, lograge is a common logging library that includes JSON formatting as well [4](https://github.com/roidrage/lograge)
* For Python, python-json-logger [5](https://pypi.org/project/python-json-logger/) is one of the many libraries for doing this

Cloud logging vendor side:

* Here's Datadog recommending the use of JSON logging as a best practice for Python when sending data to Datadog [6](https://www.datadoghq.com/blog/python-logging-best-practices/#unify-all-your-python-logs) (Disclaimer: I currently work at Datadog)
* Loggly also supports and seems to recommend JSON logging [7](https://www.loggly.com/solution/json-logging/), [8](https://www.loggly.com/use-cases/json-logging-best-practices/)
* Coralogix also supports JSON parsing [9](https://coralogix.com/docs/auto-json-parsing/)

Tools:

* Fluentd supports JSON logging [10](https://docs.fluentd.org/parser/json)
* Logstash supports it as well [11](https://www.elastic.co/guide/en/logstash/current/plugins-filters-json.html)

---

Having said that, the big downside of json logs is when you have to look at them as raw json; e.g. the json part is only useful as a payload that is easy-ish to emit and parse, but to get a good experience hopefully whatever is catching logs on the other side can parse the json and make it easier for humans to grok.

So this is why I think this @ioquatix's idea would be really cool.

**Comment by kevincox** [comment]
Newline-delimited JSON logs is definitely a popular format. Especially for Docker services. I think it would be great to be able to ingest them into the Journal native format so that it is for example easy to filter with `journalctl`.

I do think this should be opt-in for a handful of reasons:

1. Many existing services are already logging JSON and people have built automation pipelines around this. Switching silently from `MESSAGE={"foo":5,"bar":"plonk"}` to `FOO=5` and `BAR=plonk` would be a breaking change.
2. It isn't clear if journald can faithfully represent all JSON. For example there is no nesting supported in journald. So there would need to be a translation. Also upper-casing all fields will require a translation. These translations will not be lossless in the general case (unless there is some complicated escaping done which is likely worse). I think most services won't have these problems so the feature will be useful but allowing opt-in/opt-out will be critical for services that do output these formats.

**Comment by ioquatix** [comment]
For 1, I suggested `JournalFormat=json` in the service configuration file.

For 2, I don't have any strong opinion, but maybe we don't need to translate nested fields, Another option is `dot.delimited.nested.fields`.
